superset/db_engine_specs.py

```diff
@@ -1397,6 +1412,42 @@ def epoch_to_dttm(cls):
         return 'from_unixtime({col})'
 
 
+class PinotEngineSpec(BaseEngineSpec):
+    engine = 'pinot'
+    allows_subquery = False
+    inner_joins = False
+    groupby_fields_auto_present_in_select = True
+    prefer_between_for_range_filters = True
+    db_respects_given_column_labels = False
+
+    _time_grain_to_datetimeconvert = {
+        'PT1S': '1:SECONDS',
+        'PT1M': '1:MINUTES',
+        'PT1H': '1:HOURS',
+        'P1D': '1:DAYS',
+        'P1Y': '1:YEARS',
+        'P1M': '1:MONTHS',
+    }
+
+    # Pinot does its own conversion below
+    time_grain_functions = {k: None for k in _time_grain_to_datetimeconvert.keys()}
+
+    @classmethod
+    def get_time_expr(cls, expr, pdf, time_grain, grain):
+        is_epoch = pdf in ('epoch_s', 'epoch_ms')
+        if not is_epoch:
+            raise NotImplementedError('Pinot currently only supports epochs')
+        seconds_or_ms = 'MILLISECONDS' if pdf == 'epoch_ms' else 'SECONDS'
+        time_format = f'1:{seconds_or_ms}:EPOCH'
+        granularity = cls._time_grain_to_datetimeconvert.get(time_grain)
+        if not granularity:
+            raise NotImplementedError('No pinot grain spec for ' + str(time_grain))
+        # In pinot the output is a string since there is no timestamp column like pg
+        fmt = (f'DATETIMECONVERT({{col}}, \"{time_format}\", \"{time_format}\",' +
+               f'\"{granularity}\")')
+        return fmt.format(col=expr)
+
```

Comment:
This is slightly difficult to follow; a comment explaining what is happening here might be helpful.

Reply:
Okay. This UDF is weird. I will add a link to its description and some rationale.

Reply:
It's super weird, I never understood how it works.